0003-secret-seed-storage-and-key-derivation.md

```diff
@@ -0,0 +1,195 @@
+
+By default I'd put it somewhere like:
+
+```
+~/.comit/secret_seed
+```
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
+
+We should make the permissions restrictive so only the user has read permission on it.
+
+#### What about encrypting the database?
+
+I don't think it's worth encrypting the database yet until we have a motivation to do it.
+An attacker stealing the database would learn what protocol executions you're involved in so atm it's just a privacy issue.
+
+We *could* encrypt the database we the `secret_seed` seeing as they're separated.
+
+### How to generate it?
+
+If `secret_seed_file` isn't set, I think we should just generate it from `OsRng` to `~/.comit/secret_seed` if it doesn't exist.
+
+We can make a tool to generate/restore it from a mnemonic and encrypt/decrypt it later on.
+
+## Secret/NodeID Derivation
+
+I think the way we are doing it now is roughly the right way but I recommend modifying it in the following ways:
+
+### length extension attacks
+
+Informally, there are three things I think you need to ensure when doing deterministic derivation:
+
+1. The derived keys should appear to be random.
+2. Seeing a derived key shouldn't help you learn anything else about the keys that will be derived in the future (or the keys that have been derived in the past).
+3. You shouldn't use deterministic derivation where a malicious party can use this against you.
+
+[Length extension](https://en.wikipedia.org/wiki/Length_extension_attack) attacks break (2) and as we are using SHA-256 we could be susceptible to them.
+A contrived example is that let's say you derive `secret` and `secret_key` by doing `hash(seed || "secret")` and `hash(seed || "secret_key")`.
+With SHA256 it is possible to figure out what `secret_key` is once you've seen `secret` IF the length of `seed || secret` happens to be equal to the block length of the hash function (I think).
+
+Obviously, in our case you can't do this attack but I'd suggest we use a hash function that isn't susceptible to it.
+
+### Use u8 rather than strings to identify key types
+
+There's nothing inherently wrong with using utf-8 encoded strings to a hash function to produce secret keys etc. e.g. `hash(seed || "secret")` (putting length extension attacks to the side).
+This is what we do now in rfc003's [SecretSource](https://github.com/comit-network/comit-rs/blob/master/application/comit_node/src/swap_protocols/rfc003/secret_source.rs).
+There is another way to do it which is to simply encode integers for each type of key:
+
+``` rust
+enum SecretType {
+    Secret,
+    RedeemKey,
+    RefundKey,
+}
+
+..
+fn secp256k1_redeem(&self) -> KeyPair {
+    KeyPair::from_secret_key_slice(self.sha256_with_seed(&[SecretType::RedeemKey as u8]).as_ref())
```

Comment:
:x: 

`as u8` depends on the order though I guess? We should probably just assign constant values to it to make it totally refactoring safe.